clusterloader2/pkg/measurement/common/schedulerlatency.go

```diff
@@ -0,0 +1,218 @@
+}
+
+type schedulerLatencyMeasurement struct{}
+
+// Execute supports two actions:
+// - reset - Resets latency data on api server side.
+// - gather - Gathers and prints current api server latency data.
+func (*schedulerLatencyMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	var summaries []measurement.Summary
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return summaries, err
+	}
+	provider, err := util.GetString(config.Params, "provider")
+	if err != nil {
+		return summaries, err
+	}
+	host, err := util.GetStringOrDefault(config.Params, "host", "")
+	if err != nil {
+		return summaries, err
+	}
+	masterName, err := util.GetStringOrDefault(config.Params, "masterName", "")
+	if err != nil {
+		return summaries, err
+	}
+
+	switch action {
+	case "reset":
+		glog.Infof("Resetting latency metrics in schduler...")
+		return summaries, resetSchedulerMetrics(config.ClientSet, provider, host, masterName)
+	case "gather":
+		return getSchedulingLatency(config.ClientSet, provider, host, masterName)
+	default:
+		return summaries, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+func resetSchedulerMetrics(c clientset.Interface, provider, host, masterName string) error {
+	_, err := sendRestRequestToScheduler(c, "DELETE", provider, host, masterName)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Retrieves scheduler latency metrics.
+func getSchedulingLatency(c clientset.Interface, provider, host, masterName string) ([]measurement.Summary, error) {
+	var summaries []measurement.Summary
+	result := schedulingMetrics{}
+	data, err := sendRestRequestToScheduler(c, "GET", provider, host, masterName)
+	if err != nil {
+		return summaries, err
+	}
+
+	samples, err := measurementutil.ExtractMetricSamples(data)
+	if err != nil {
+		return summaries, err
+	}
+
+	for _, sample := range samples {
+		if sample.Metric[model.MetricNameLabel] != schedulingLatencyMetricName {
+			continue
+		}
+
+		var metric *measurementutil.LatencyMetric
+		switch sample.Metric[schedulermetric.OperationLabel] {
+		case schedulermetric.PredicateEvaluation:
+			metric = &result.PredicateEvaluationLatency
+		case schedulermetric.PriorityEvaluation:
+			metric = &result.PriorityEvaluationLatency
+		case schedulermetric.PreemptionEvaluation:
+			metric = &result.PreemptionEvaluationLatency
+		case schedulermetric.Binding:
+			metric = &result.BindingLatency
+		}
+		if metric == nil {
+			continue
+		}
+
+		quantile, err := strconv.ParseFloat(string(sample.Metric[model.QuantileLabel]), 64)
+		if err != nil {
+			return summaries, err
+		}
+		metric.SetQuantile(quantile, time.Duration(int64(float64(sample.Value)*float64(time.Second))))
+	}
+	summaries = append(summaries, &result)
+
+	return summaries, nil
+}
+
+// Sends REST request to kube scheduler metrics
+func sendRestRequestToScheduler(c clientset.Interface, op, provider, host, masterName string) (string, error) {
```

Comment:
Btw - even though you probably copied over this file - I think you should name this function sth better than sendRestRequest.. (as that's only one of the case, the other being SSH).

Reply:
>  I think you should name this function sth better than sendRestRequest.. (as that's only one of the case, the other being SSH).

I will not agree with the statement that this function does something more than sending request.
In case of this ssh, ssh is here only used to execute `curl -X` which is sending the request. As you can see, main point of this function is sending request, with accessing component or sshing to master being only a way to achieve this.

Reply:
I'm not saying it's not a request. I meant about the 'Rest' part (which is only one of the cases - the other case SSH is not a rest request).

Reply:
I still think the name should be more clearer, like `getRawSchedulerMetrics` or sth. The current name doesn't say what kind of request it is.

Reply:
> The current name doesn't say what kind of request it is.

It takes request kind as parameter `op`.
`getRawSchedulerMetrics` - this function doesn't always get raw scheduler metrics. If you send `DELETE`, you will reset metrics instead.

Reply:
Ok, then sgtm.